### DISCUSSION TITLE
Redis questions please

### DISCUSSION THREAD

**Discussion by timkgh** [rspamd/rspamd#4266] [Q&A]
1. Is Redis required for a functioning rspamd? If not, what functionality is lost when not using Redis?
2. For a very low volume personal email server (100-200 emails/day), what memory requirements does rspamd with Redis have? Would it work on a 1GB RAM VPS instance that is dedicated to email only and limit Redis to 500MB memory or even less?
3. In a Redis master-slave setup (not using Sentinel) where the secondary MX reads from the local slave Redis but writes to the master Redis on the primary MX, what happens when the Master Redis goes down? Will email delivery stop working completely on the primary MX? What about the secondary MX, it can continue reading from the slave Redis but cannot write to Master anymore, will email delivery just stop or will it continue to work with rspamd?
4. The documentation suggests setting up 3 Redis instances and only synchronizing master-slave for the bayes and fuzzy instances. What happens if one uses a single Redis instance due to memory limitations and syncs the entire Redis instance? I'm not looking for perfection on the secondary MX, I just don't want it to be completely broken.
5. In general, in a Postfix + Rspamd setup, if rspamd is down, what happens to email delivery? Will Postfix continue to deliver emails without any spam filtering, will it temporarily reject email so the sender will retry? Will it put it in a local queue (deferred or hold queue?) and retry delivery later?

**Comment by tokred** [comment]
Trying a late answer here, no dev though, only operator of a small, low traffic but redundant personal mail env. I run two MX with replicated Redis, incl. Sentinel on those + on a third host as quorum (i.e. 2 replicas, 3 sentinels):

1. I'd say yes, and not worth running rspamd without Redis as many modules/features rely on it.
2. In my environment, Redis has a very low memory footprint, barely noticeable. I limit `maxmemory` to 128mb, runs perfectly fine on two small VPS with other stuff like Postfix, Dovecot, etc (ClamAV is the real memory hog on it). The on-disk dump file for me is <10mb, while used RAM for Redis is <100mb.
3. No guarantee but I had a Redis cluster issue because of a stupid error and the master was unavailable, and email delivery continued just fine while rspamd did not really care and simply not wrote any new state data. It logged some errors but continued to read data and process emails. Primary/secondary MX did not make a difference, but unaffected.
4. Frankly, I read this quite a while ago and decided to disagree: In my opinion, you definitely want state data synced between multiple MX too, and not only bayes/fuzzy (i.e. also things like replies, greylisting, etc!) but of course also DKIM/ARC keys (if in Redis). I also decided against splitting instances and I am using a single replicated instance perfectly fine, you'd only have to make a decision regarding [key eviction](https://redis.io/docs/reference/eviction/) for the (unlikely) case of hitting `maxmemory`: By default, Redis would stop writing new data which is likely not what you want. The rspamd docs explicitly mention `volatile-ttl` for Bayes instance only, but in my opinion it is also a safe choice for a single instance. I think this is much more relevant for larger environments where you will likely hit `maxmemory` repeatedly and need Redis to clean up aggressively.
5. Postfix has the `milter_default_action` setting which I have set to non-default `accept`, to address this exact scenario. See the [Postfix docs](https://www.postfix.org/MILTER_README.html#errors), Postfix will then deliver emails without any filtering.
